# Pattern Mining - Association Rule Mining

A *frequent pattern* is a substructure that appears frequently in a dataset. Finding the frequent patterns of a dataset is a essential step in data mining tasks such as *feature extraction* and a necessary ingredient of *association rule learning*. This kind of algorithms are extremely useful in the field of *Market Basket Analysis*, which in turn provide retailers with invaluable information about their customer shopping habbits and needs. 

Here, I will shortly describe the *GLC frequent pattern mining toolkit*, the tools it provides and its functionality. Major advantage of this high-level ML toolkit is the ease it provides to train an association rule mining algorithm, as well as the high interpretability of the returned results. Under the hood the GLG frequent pattern mining toolkit runs a *[TFP-Growth algorithm](https://www.computer.org/csdl/trans/tk/2005/05/k0652-abs.html)*, introduced by Wang, Jianyong, et al. in 2005. For a recent review of the various directions in the field consult *[Han, Jiawei, et al. "Frequent pattern mining: current status and future directions.", Data Mining and Knowledge Discovery 15.1 (2007): 55-86](http://link.springer.com/article/10.1007%2Fs10618-006-0059-1)*.

## Fire up GraphLab Create

In [1]:
import graphlab as gl
from graphlab import aggregate as agg

## A simple retailer example

Here we discuss a simple example of receipt data from a bakery. The dataset consists of items like `'ApplePie'` and `'GanacheCookie'`. The task is to identify sets of items that are frequently bought together. The dataset consists of *266209 rows* and *6 columns* which look like the following. The dataset was constructed by modifying the *[Extended BAKERY dataset](https://wiki.csc.calpoly.edu/datasets/wiki/ExtendedBakery)*.

In [2]:
bakery_sf = gl.SFrame('./bakery_sf')
bakery_sf

2016-05-09 01:13:58,787 [INFO] graphlab.cython.cy_server, 176: GraphLab Create v1.9 started. Logging: /tmp/graphlab_server_1462745636.log


This non-commercial license of GraphLab Create is assigned to tgrammat@gmail.com and will expire on September 21, 2016. For commercial licensing options, visit https://dato.com/buy/.


Columns:
	Receipt	int
	SaleDate	str
	EmpId	int
	StoreNum	int
	Quantity	int
	Item	str

Rows: 266209

Data:
+---------+-------------+-------+----------+----------+-----------------+
| Receipt |   SaleDate  | EmpId | StoreNum | Quantity |       Item      |
+---------+-------------+-------+----------+----------+-----------------+
|    1    | 12-JAN-2000 |   20  |    20    |    1     |  GanacheCookie  |
|    1    | 12-JAN-2000 |   20  |    20    |    5     |     ApplePie    |
|    2    | 15-JAN-2000 |   35  |    10    |    1     |   CoffeeEclair  |
|    2    | 15-JAN-2000 |   35  |    10    |    3     |     ApplePie    |
|    2    | 15-JAN-2000 |   35  |    10    |    4     |   AlmondTwist   |
|    2    | 15-JAN-2000 |   35  |    10    |    3     |    HotCoffee    |
|    3    |  8-JAN-2000 |   13  |    13    |    5     |    OperaCake    |
|    3    |  8-JAN-2000 |   13  |    13    |    3     |   OrangeJuice   |
|    3    |  8-JAN-2000 |   13  |    13    |    3     | CheeseCroissant |
|    4    | 24-JAN-2000 |   16  |    16    |    1     |   TruffleCake   |
+---------+-------------+-------+----------+----------+-----------------+
[266209 rows x 6 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

As we can see below, all the coffee products have similar sale frequencies and there is no some particular subset of products that is more preferred than the remaining ones.

In [3]:
gl.canvas.set_target('ipynb')
bakery_sf['Item'].show()

Next, we split the **`bakery_sf`** data set in a **training** and a **test part**.

In [4]:
(train, test) = bakery_sf.random_split(0.8, seed=1)

In [5]:
print 'Number of Rows in training set [80pct of Known Examples]: %d'   % train.num_rows()
print 'Number of Rows in test set [20pct of Known Examples]: %d'   % test.num_rows()

Number of Rows in training set [80pct of Known Examples]: 212930
Number of Rows in test set [20pct of Known Examples]: 53279


In order to run a frequent pattern mining algorithm, we require an **item column**, (the column **`'Item'`** in this example), and a set of **feature columns** that *uniquely identify* a transaction (the columns **`['Receipt', 'StoreNum']`** in this example, since we need to take in account the geophraphical location of each store and the accompanied social-economic criteria that may exist).

In addition we need to specify the 3 basic parameters of the **FP-Growth algorithm** which is called by the high-level GraphLab Create (GLC) function. These are:

* **`min_support`:** The minimum number of times that a pattern must occur in order to be considered a *frequent* one. Here, we choose a threshold of 1&permil; of total transactions in record to be the `min_support`.
* **`max_patterns`:** The maximum number of frequent patterns to be mined.
* **`min_length`:** The minimum size (number of elements in the set) of each pattern being mined.

In [6]:
min_support = int(train.num_rows()*0.001)
model = gl.frequent_pattern_mining.create(train, 'Item', features=['Receipt', 'StoreNum'], 
                                          min_support=min_support, max_patterns=500, min_length=4)

Indexing complete. Found 50 unique items.

Preprocessing complete. Found 73566 unique transactions.

Building frequent pattern tree.

+-----------+----------------+------------+----------------------+------------------+

| Iteration | Num. Patterns  | Support    | Current Min Support  | Elapsed Time     |

+-----------+----------------+------------+----------------------+------------------+

| 0         | 0              | 6550       | 212                  | 153us            |

| 1         | 0              | 6167       | 212                  | 216us            |

| 2         | 0              | 6050       | 212                  | 258us            |

| 3         | 0              | 5587       | 212                  | 304us            |

| 4         | 0              | 5563       | 212                  | 352us            |

| 5         | 0              | 5548       | 212                  | 426us            |

| 6         | 0              | 5495       | 212                  | 520us            |

| 7         | 0              | 5438       | 212                  | 625us            |

| 8         | 0              | 5376       | 212                  | 774us            |

| 9         | 0              | 5088       | 212                  | 983us            |

| 10        | 0              | 5071       | 212                  | 1.152ms          |

| 11        | 0              | 5040       | 212                  | 1.378ms          |

| 12        | 0              | 4994       | 212                  | 1.775ms          |

| 13        | 0              | 4981       | 212                  | 2.033ms          |

| 14        | 0              | 4949       | 212                  | 2.323ms          |

| 15        | 0              | 4948       | 212                  | 2.688ms          |

| 16        | 0              | 4932       | 212                  | 3.178ms          |

| 17        | 0              | 4914       | 212                  | 7.875ms          |

| 18        | 0              | 4913       | 212                  | 8.852ms          |

| 19        | 0              | 4690       | 212                  | 9.647ms          |

| 20        | 0              | 4681       | 212                  | 10.335ms         |

| 21        | 0              | 4613       | 212                  | 11.093ms         |

| 22        | 0              | 4592       | 212                  | 15.207ms         |

| 23        | 1              | 4568       | 212                  | 18.231ms         |

| 24        | 1              | 4506       | 212                  | 20.094ms         |

| 25        | 1              | 4496       | 212                  | 21.479ms         |

| 26        | 1              | 4398       | 212                  | 22.568ms         |

| 27        | 1              | 4122       | 212                  | 26.167ms         |

| 28        | 1              | 4101       | 212                  | 27.718ms         |

| 29        | 1              | 4074       | 212                  | 28.808ms         |

| 30        | 1              | 4071       | 212                  | 29.861ms         |

| 31        | 1              | 4057       | 212                  | 30.824ms         |

| 32        | 1              | 4056       | 212                  | 31.828ms         |

| 33        | 1              | 4055       | 212                  | 33.113ms         |

| 34        | 1              | 4042       | 212                  | 38.03ms          |

| 35        | 2              | 4014       | 212                  | 40.764ms         |

| 36        | 2              | 3711       | 212                  | 42.374ms         |

| 37        | 3              | 3707       | 212                  | 43.949ms         |

| 38        | 8              | 2661       | 212                  | 48.01ms          |

| 39        | 8              | 2619       | 212                  | 51.163ms         |

| 40        | 8              | 2595       | 212                  | 52.532ms         |

| 41        | 8              | 2593       | 212                  | 53.917ms         |

| 42        | 8              | 2576       | 212                  | 55.296ms         |

| 43        | 8              | 2551       | 212                  | 60.564ms         |

| 44        | 8              | 2544       | 212                  | 62.095ms         |

| 45        | 8              | 2538       | 212                  | 66.237ms         |

| 46        | 8              | 2533       | 212                  | 68.018ms         |

| 47        | 8              | 2530       | 212                  | 69.937ms         |

| 48        | 8              | 2521       | 212                  | 71.781ms         |

| 49        | 8              | 2511       | 212                  | 77.793ms         |

| Final     | 8              | -          | 212                  | 79.672ms         |

+-----------+----------------+------------+----------------------+------------------+

Pattern mining complete. Found 8 unique closed patterns.

Here, we obtain the most frequent feature patterns.

In [7]:
print 'The most frequent feature patters are:'
print '-----------------------------------------'
model.frequent_patterns.print_rows(max_column_width=70)

The most frequent feature patters are:
-----------------------------------------
+-----------------------------------------------------------------------+---------+
|                                pattern                                | support |
+-----------------------------------------------------------------------+---------+
|            [CoffeeEclair, HotCoffee, ApplePie, AlmondTwist]           |   877   |
|    [LemonLemonade, RaspberryCookie, LemonCookie, RaspberryLemonade]   |   813   |
|        [LemonLemonade, RaspberryCookie, LemonCookie, GreenTea]        |   649   |
|      [RaspberryCookie, LemonCookie, RaspberryLemonade, GreenTea]      |   645   |
|          [AppleTart, AppleDanish, AppleCroissant, CherrySoda]         |   640   |
|       [LemonLemonade, LemonCookie, RaspberryLemonade, GreenTea]       |   630   |
|     [LemonLemonade, RaspberryCookie, RaspberryLemonade, GreenTea]     |   620   |
| [LemonLemonade, RaspberryCookie, LemonCookie, RaspberryLemonade, G... |   512

Note that the **`'pattern'`** column contains the patterns that occur frequently together, whereas the **`'support'`** column contains the number of times these patterns occur together in the entire dataset. 

In this example, the pattern:
```
[CoffeeEclair, HotCoffee, ApplePie, AlmondTwist]
```
occurred 877 times in the training data.


**Definition**

> A **frequent pattern** is a *set of items* with a *support greater than* user-specified **minimum support threshold**.

However, there is significant redundancy in mining frequent patterns; every subset of a frequent pattern is also frequent (e.g. `'CoffeeEclair'` must be frequent if `['CoffeeEclair'`, `'HotCoffee']` is frequent). The frequent pattern mining toolkit avoids this redundancy by mining the closed frequent patterns, i.e. *frequent patterns with no superset of the same support*. This is achieved by the very design of the **FP-Growth Algorithm**.

Note, that by relaxing the **`min_length`** requirement, one can obtain more frequent patterns of sold coffe products.

In [8]:
min_support = int(train.num_rows()*0.001)
model = gl.frequent_pattern_mining.create(train, 'Item', features=['Receipt', 'StoreNum'], 
                                          min_support=min_support, max_patterns=500, min_length=3)

Indexing complete. Found 50 unique items.

Preprocessing complete. Found 73566 unique transactions.

Building frequent pattern tree.

+-----------+----------------+------------+----------------------+------------------+

| Iteration | Num. Patterns  | Support    | Current Min Support  | Elapsed Time     |

+-----------+----------------+------------+----------------------+------------------+

| 0         | 0              | 6550       | 212                  | 148us            |

| 1         | 0              | 6167       | 212                  | 205us            |

| 2         | 0              | 6050       | 212                  | 246us            |

| 3         | 0              | 5587       | 212                  | 291us            |

| 4         | 0              | 5563       | 212                  | 338us            |

| 5         | 0              | 5548       | 212                  | 411us            |

| 6         | 0              | 5495       | 212                  | 491us            |

| 7         | 0              | 5438       | 212                  | 593us            |

| 8         | 0              | 5376       | 212                  | 741us            |

| 9         | 0              | 5088       | 212                  | 963us            |

| 10        | 0              | 5071       | 212                  | 1.114ms          |

| 11        | 0              | 5040       | 212                  | 1.344ms          |

| 12        | 0              | 4994       | 212                  | 1.733ms          |

| 13        | 0              | 4981       | 212                  | 2.001ms          |

| 14        | 0              | 4949       | 212                  | 2.298ms          |

| 15        | 1              | 4948       | 212                  | 2.664ms          |

| 16        | 2              | 4932       | 212                  | 3.176ms          |

| 17        | 2              | 4914       | 212                  | 9.403ms          |

| 18        | 2              | 4913       | 212                  | 10.19ms          |

| 19        | 2              | 4690       | 212                  | 10.995ms         |

| 20        | 3              | 4681       | 212                  | 11.731ms         |

| 21        | 3              | 4613       | 212                  | 12.517ms         |

| 22        | 3              | 4592       | 212                  | 13.692ms         |

| 23        | 7              | 4568       | 212                  | 14.531ms         |

| 24        | 7              | 4506       | 212                  | 19.159ms         |

| 25        | 7              | 4496       | 212                  | 20.558ms         |

| 26        | 8              | 4398       | 212                  | 22.477ms         |

| 27        | 8              | 4122       | 212                  | 23.717ms         |

| 28        | 8              | 4101       | 212                  | 26.617ms         |

| 29        | 9              | 4074       | 212                  | 28.513ms         |

| 30        | 9              | 4071       | 212                  | 29.498ms         |

| 31        | 9              | 4057       | 212                  | 30.41ms          |

| 32        | 9              | 4056       | 212                  | 31.413ms         |

| 33        | 10             | 4055       | 212                  | 32.858ms         |

| 34        | 11             | 4042       | 212                  | 36.599ms         |

| 35        | 15             | 4014       | 212                  | 39.82ms          |

| 36        | 16             | 3711       | 212                  | 41.412ms         |

| 37        | 20             | 3707       | 212                  | 42.989ms         |

| 38        | 31             | 2661       | 212                  | 48.438ms         |

| 39        | 31             | 2619       | 212                  | 50.493ms         |

| 40        | 31             | 2595       | 212                  | 51.86ms          |

| 41        | 31             | 2593       | 212                  | 53.233ms         |

| 42        | 31             | 2576       | 212                  | 54.621ms         |

| 43        | 31             | 2551       | 212                  | 60.232ms         |

| 44        | 31             | 2544       | 212                  | 61.927ms         |

| 45        | 31             | 2538       | 212                  | 63.684ms         |

| 46        | 31             | 2533       | 212                  | 67.747ms         |

| 47        | 31             | 2530       | 212                  | 69.606ms         |

| 48        | 31             | 2521       | 212                  | 71.269ms         |

| 49        | 31             | 2511       | 212                  | 72.901ms         |

| Final     | 31             | -          | 212                  | 77.761ms         |

+-----------+----------------+------------+----------------------+------------------+

Pattern mining complete. Found 31 unique closed patterns.

In [9]:
print 'The most frequent feature patters are:'
print '-----------------------------------------'
model.frequent_patterns.print_rows(num_rows=35, max_column_width=70)

The most frequent feature patters are:
-----------------------------------------
+-----------------------------------------------------------------------+---------+
|                                pattern                                | support |
+-----------------------------------------------------------------------+---------+
|                 [CherryTart, ApricotDanish, OperaCake]                |   1561  |
|                 [CoffeeEclair, ApplePie, AlmondTwist]                 |   1351  |
|              [ChocolateCake, ChocolateCoffee, CasinoCake]             |   1308  |
|              [HotCoffee, ApricotCroissant, BlueberryTart]             |   1220  |
|                  [CoffeeEclair, HotCoffee, ApplePie]                  |   1109  |
|                   [HotCoffee, ApplePie, AlmondTwist]                  |   1100  |
|                 [CoffeeEclair, HotCoffee, AlmondTwist]                |   1086  |
|             [CoffeeEclair, BlackberryTart, SingleEspresso]            |   102

Alternatively, by decreasing the **`min_support`** one can obtain more patterns of sold coffee products which are again assumed frequent but with respect to this new threshold.

In [10]:
min_support = int(train.num_rows()*(1e-04))
model = gl.frequent_pattern_mining.create(train, 'Item', features=['Receipt', 'StoreNum'], 
                                          min_support=min_support, max_patterns=500, min_length=4)

Indexing complete. Found 50 unique items.

Preprocessing complete. Found 73566 unique transactions.

Building frequent pattern tree.

+-----------+----------------+------------+----------------------+------------------+

| Iteration | Num. Patterns  | Support    | Current Min Support  | Elapsed Time     |

+-----------+----------------+------------+----------------------+------------------+

| 0         | 0              | 6550       | 21                   | 139us            |

| 1         | 0              | 6167       | 21                   | 197us            |

| 2         | 0              | 6050       | 21                   | 238us            |

| 3         | 0              | 5587       | 21                   | 282us            |

| 4         | 0              | 5563       | 21                   | 337us            |

| 5         | 0              | 5548       | 21                   | 408us            |

| 6         | 0              | 5495       | 21                   | 527us            |

| 7         | 0              | 5438       | 21                   | 657us            |

| 8         | 0              | 5376       | 21                   | 815us            |

| 9         | 0              | 5088       | 21                   | 1.09ms           |

| 10        | 0              | 5071       | 21                   | 1.506ms          |

| 11        | 0              | 5040       | 21                   | 1.808ms          |

| 12        | 0              | 4994       | 21                   | 2.307ms          |

| 13        | 0              | 4981       | 21                   | 2.872ms          |

| 14        | 0              | 4949       | 21                   | 3.658ms          |

| 15        | 7              | 4948       | 21                   | 6.375ms          |

| 16        | 10             | 4932       | 21                   | 7.598ms          |

| 17        | 11             | 4914       | 21                   | 11.079ms         |

| 18        | 12             | 4913       | 21                   | 14.42ms          |

| 19        | 13             | 4690       | 21                   | 19.791ms         |

| 20        | 14             | 4681       | 21                   | 23.49ms          |

| 21        | 15             | 4613       | 21                   | 27.218ms         |

| 22        | 15             | 4592       | 21                   | 33.589ms         |

| 23        | 16             | 4568       | 21                   | 41.8ms           |

| 24        | 17             | 4506       | 21                   | 47.385ms         |

| 25        | 18             | 4496       | 21                   | 50.504ms         |

| 26        | 25             | 4398       | 21                   | 62.404ms         |

| 27        | 26             | 4122       | 21                   | 68.23ms          |

| 28        | 28             | 4101       | 21                   | 73.959ms         |

| 29        | 28             | 4074       | 21                   | 79.732ms         |

| 30        | 28             | 4071       | 21                   | 85.089ms         |

| 31        | 29             | 4057       | 21                   | 90.266ms         |

| 32        | 31             | 4056       | 21                   | 96.116ms         |

| 33        | 33             | 4055       | 21                   | 101.783ms        |

| 34        | 35             | 4042       | 21                   | 108.571ms        |

| 35        | 37             | 4014       | 21                   | 116.118ms        |

| 36        | 40             | 3711       | 21                   | 125.313ms        |

| 37        | 42             | 3707       | 21                   | 133.405ms        |

| 38        | 47             | 2661       | 21                   | 141.554ms        |

| 39        | 49             | 2619       | 21                   | 150.142ms        |

| 40        | 51             | 2595       | 21                   | 158.276ms        |

| 41        | 51             | 2593       | 21                   | 167.16ms         |

| 42        | 51             | 2576       | 21                   | 175.629ms        |

| 43        | 52             | 2551       | 21                   | 184.835ms        |

| 44        | 53             | 2544       | 21                   | 194.06ms         |

| 45        | 55             | 2538       | 21                   | 204.208ms        |

| 46        | 56             | 2533       | 21                   | 214.367ms        |

| 47        | 56             | 2530       | 21                   | 224.494ms        |

| 48        | 57             | 2521       | 21                   | 234.381ms        |

| 49        | 58             | 2511       | 21                   | 244.784ms        |

| Final     | 59             | -          | 21                   | 255.337ms        |

+-----------+----------------+------------+----------------------+------------------+

Pattern mining complete. Found 59 unique closed patterns.

In [11]:
print 'The most frequent feature patters are:'
print '-----------------------------------------'
model.frequent_patterns.print_rows(num_rows=60, max_column_width=70)

The most frequent feature patters are:
-----------------------------------------
+-----------------------------------------------------------------------+---------+
|                                pattern                                | support |
+-----------------------------------------------------------------------+---------+
|            [CoffeeEclair, HotCoffee, ApplePie, AlmondTwist]           |   877   |
|    [LemonLemonade, RaspberryCookie, LemonCookie, RaspberryLemonade]   |   813   |
|        [LemonLemonade, RaspberryCookie, LemonCookie, GreenTea]        |   649   |
|      [RaspberryCookie, LemonCookie, RaspberryLemonade, GreenTea]      |   645   |
|          [AppleTart, AppleDanish, AppleCroissant, CherrySoda]         |   640   |
|       [LemonLemonade, LemonCookie, RaspberryLemonade, GreenTea]       |   630   |
|     [LemonLemonade, RaspberryCookie, RaspberryLemonade, GreenTea]     |   620   |
| [LemonLemonade, RaspberryCookie, LemonCookie, RaspberryLemonade, G... |   512

To see some details of the trained model:

In [12]:
print model

Class                         : FrequentPatternMiner

Model fields
------------
Min support                   : 21
Max patterns                  : 500
Min pattern length            : 4

Most frequent patterns
----------------------
['CoffeeEclair', 'HotCoffee', 'ApplePie', 'AlmondTwist']: 877
['LemonLemonade', 'RaspberryCookie', 'LemonCookie', 'RaspberryLemonade']: 813
['LemonLemonade', 'RaspberryCookie', 'LemonCookie', 'GreenTea']: 649
['RaspberryCookie', 'LemonCookie', 'RaspberryLemonade', 'GreenTea']: 645
['AppleTart', 'AppleDanish', 'AppleCroissant', 'CherrySoda']: 640
['LemonLemonade', 'LemonCookie', 'RaspberryLemonade', 'GreenTea']: 630
['LemonLemonade', 'RaspberryCookie', 'RaspberryLemonade', 'GreenTea']: 620
['LemonLemonade', 'RaspberryCookie', 'LemonCookie', 'RaspberryLemonade', 'GreenTea']: 512
['HotCoffee', 'CherryTart', 'ApricotDanish', 'OperaCake']: 29
['CherryTart', 'ApricotDanish', 'OperaCake', 'AlmondBear Claw']: 29



## Top-k frequent patterns

In practice, we rarely know the appropriate **`min_support`** threshold to use. As an alternative to specifying a minimum support, we can specify a **maximum number of patterns** to mine using the **`max_patterns`** parameter. Instead of mining all patterns above a minimum support threshold, we mine the most frequent patterns until the maximum number of closed patterns are round. For large data sets, this mining process can be time-consuming. We recommend specifying a somehow *large initial minimum support* bound to speed up the mining.

In [13]:
min_support = int(train.num_rows()*1e-03)
top5_freq_patterns = gl.frequent_pattern_mining.create(train, 'Item',
                                                       features=['Receipt', 'StoreNum'],
                                                       min_support=min_support,
                                                       max_patterns=5,
                                                       min_length=4)

Indexing complete. Found 50 unique items.

Preprocessing complete. Found 73566 unique transactions.

Building frequent pattern tree.

+-----------+----------------+------------+----------------------+------------------+

| Iteration | Num. Patterns  | Support    | Current Min Support  | Elapsed Time     |

+-----------+----------------+------------+----------------------+------------------+

| 0         | 0              | 6550       | 212                  | 126us            |

| 1         | 0              | 6167       | 212                  | 184us            |

| 2         | 0              | 6050       | 212                  | 224us            |

| 3         | 0              | 5587       | 212                  | 268us            |

| 4         | 0              | 5563       | 212                  | 313us            |

| 5         | 0              | 5548       | 212                  | 383us            |

| 6         | 0              | 5495       | 212                  | 458us            |

| 7         | 0              | 5438       | 212                  | 557us            |

| 8         | 0              | 5376       | 212                  | 698us            |

| 9         | 0              | 5088       | 212                  | 916us            |

| 10        | 0              | 5071       | 212                  | 1.061ms          |

| 11        | 0              | 5040       | 212                  | 1.289ms          |

| 12        | 0              | 4994       | 212                  | 1.667ms          |

| 13        | 0              | 4981       | 212                  | 1.927ms          |

| 14        | 0              | 4949       | 212                  | 2.214ms          |

| 15        | 0              | 4948       | 212                  | 2.563ms          |

| 16        | 0              | 4932       | 212                  | 3.054ms          |

| 17        | 0              | 4914       | 212                  | 3.82ms           |

| 18        | 0              | 4913       | 212                  | 4.456ms          |

| 19        | 0              | 4690       | 212                  | 5.202ms          |

| 20        | 0              | 4681       | 212                  | 5.872ms          |

| 21        | 0              | 4613       | 212                  | 6.583ms          |

| 22        | 0              | 4592       | 212                  | 7.688ms          |

| 23        | 1              | 4568       | 212                  | 8.467ms          |

| 24        | 1              | 4506       | 212                  | 9.335ms          |

| 25        | 1              | 4496       | 212                  | 10.57ms          |

| 26        | 1              | 4398       | 212                  | 11.645ms         |

| 27        | 1              | 4122       | 212                  | 12.801ms         |

| 28        | 1              | 4101       | 212                  | 13.625ms         |

| 29        | 1              | 4074       | 212                  | 14.626ms         |

| 30        | 1              | 4071       | 212                  | 15.654ms         |

| 31        | 1              | 4057       | 212                  | 16.642ms         |

| 32        | 1              | 4056       | 212                  | 17.735ms         |

| 33        | 1              | 4055       | 212                  | 19.332ms         |

| 34        | 1              | 4042       | 212                  | 20.706ms         |

| 35        | 2              | 4014       | 212                  | 22.113ms         |

| 36        | 2              | 3711       | 212                  | 23.685ms         |

| 37        | 3              | 3707       | 212                  | 25.331ms         |

| 38        | 5              | 2661       | 640                  | 27.118ms         |

| 39        | 5              | 2619       | 640                  | 28.411ms         |

| 40        | 5              | 2595       | 640                  | 29.833ms         |

| 41        | 5              | 2593       | 640                  | 31.224ms         |

| 42        | 5              | 2576       | 640                  | 32.85ms          |

| 43        | 5              | 2551       | 640                  | 34.47ms          |

| 44        | 5              | 2544       | 640                  | 36.051ms         |

| 45        | 5              | 2538       | 640                  | 37.627ms         |

| 46        | 5              | 2533       | 640                  | 39.176ms         |

| 47        | 5              | 2530       | 640                  | 40.797ms         |

| 48        | 5              | 2521       | 640                  | 42.609ms         |

| 49        | 5              | 2511       | 640                  | 44.434ms         |

| Final     | 5              | -          | 640                  | 46.272ms         |

+-----------+----------------+------------+----------------------+------------------+

Pattern mining complete. Found 5 unique closed patterns.

In [14]:
print top5_freq_patterns

Class                         : FrequentPatternMiner

Model fields
------------
Min support                   : 212
Max patterns                  : 5
Min pattern length            : 4

Most frequent patterns
----------------------
['CoffeeEclair', 'HotCoffee', 'ApplePie', 'AlmondTwist']: 877
['LemonLemonade', 'RaspberryCookie', 'LemonCookie', 'RaspberryLemonade']: 813
['LemonLemonade', 'RaspberryCookie', 'LemonCookie', 'GreenTea']: 649
['RaspberryCookie', 'LemonCookie', 'RaspberryLemonade', 'GreenTea']: 645
['AppleTart', 'AppleDanish', 'AppleCroissant', 'CherrySoda']: 640



We can always save the trained model by calling:

In [15]:
top5_freq_patterns.save('top5_freq_patterns_model')

### Business Application: Compute Association Rules and Make Predictions

An association rule is an ordered pair of item sets (prefix $A$, prediction $B$) denoted $A\Rightarrow B$ such that A and B are disjoint whereas $A\cup B$ is frequent. The most popular criteria for scoring association rules is to measure the **confidence of the rule**: *the ratio of the support of $A\cup B$ to the support of $A$*.

$$
Confidence(A\Rightarrow B) = \fraq{\textrm{Supp}(A\cup B)}{\textrm{Supp}(A)}.
$$

The confidence of the rule $A\Rightarrow B$ is our empirical estimate of the conditional probability for $B$ given $A$.

One can make predictions using the **`predict()`** or **`predict_topk()`** method for single and multiple predictions respectively. The output of both the methods is an **SFrame** with the following columns:

* **prefix**: The antecedent or left-hand side of an association rule. It must be a frequent pattern and a subset of the associated pattern.
* **prediction**: The consequent or right-hand side of the association rule. It must be disjoint of the prefix.
* **confidence**: The confidence of the association rule as defined above.
* **prefix support**: The frequency of the prefix pattern in the training data.
* **joint support**: The frequency of the co-occurrence ( prefix + prediction) in the training data

In [16]:
predictions = top5_freq_patterns.predict(test)

Preprocessing complete. Found 39293 unique transactions.

In [17]:
predictions.print_rows(max_column_width=100)

+---------+----------+----------------------------+--------------------+-----------------+----------------+
| Receipt | StoreNum |           prefix           |     prediction     |    confidence   | prefix support |
+---------+----------+----------------------------+--------------------+-----------------+----------------+
|  35020  |    14    |             []             |   [CoffeeEclair]   | 0.0890356958378 |     73566      |
|   6447  |    5     | [BlueberryTart, HotCoffee] | [ApricotCroissant] |  0.728793309438 |      1674      |
|  12900  |    10    |      [BlueberryTart]       | [ApricotCroissant] |  0.412691996766 |      4948      |
|  49658  |    20    |      [BlueberryTart]       | [ApricotCroissant] |  0.412691996766 |      4948      |
|  20227  |    19    |    [RaspberryLemonade]     | [RaspberryCookie]  |  0.334982681841 |      4042      |
|  30891  |    12    |      [BlueberryTart]       | [ApricotCroissant] |  0.412691996766 |      4948      |
|  47850  |    2     |      

## Feature Engineering to help other ML Tasks in Pipeline 

### The `.extract_feature()` method

Using the set of closed patterns, we can convert pattern data to binary features vectors. These feature vectors can be used for other machine learning tasks, such as **clustering** or **classification**. For each input pattern `x`, the `j-th` extracted feature `f_{x}[j]` is a binary indicator of whether the `j-th` closed pattern is contained in `x`.

First, we train the **`top100_freq_patterns`** model as shown below:

In [18]:
top100_freq_patterns = gl.frequent_pattern_mining.create(train, 'Item',
                                                         features=['Receipt', 'StoreNum'],
                                                         min_support=1, # occurs at least once in our data record
                                                         max_patterns = 100, # do not search for more than 100 patterns
                                                         min_length=2) # test data have only one coffee product sold per tid . We search for patterns of at least 2 coffee products

Indexing complete. Found 50 unique items.

Preprocessing complete. Found 73566 unique transactions.

Building frequent pattern tree.

+-----------+----------------+------------+----------------------+------------------+

| Iteration | Num. Patterns  | Support    | Current Min Support  | Elapsed Time     |

+-----------+----------------+------------+----------------------+------------------+

| 0         | 0              | 6550       | 126                  | 128us            |

| 1         | 0              | 6167       | 126                  | 183us            |

| 2         | 1              | 6050       | 126                  | 226us            |

| 3         | 3              | 5587       | 126                  | 272us            |

| 4         | 6              | 5563       | 126                  | 439us            |

| 5         | 10             | 5548       | 126                  | 522us            |

| 6         | 15             | 5495       | 126                  | 628us            |

| 7         | 21             | 5438       | 126                  | 765us            |

| 8         | 28             | 5376       | 126                  | 932us            |

| 9         | 36             | 5088       | 126                  | 10.29ms          |

| 10        | 45             | 5071       | 126                  | 11.468ms         |

| 11        | 55             | 5040       | 126                  | 11.843ms         |

| 12        | 66             | 4994       | 126                  | 12.698ms         |

| 13        | 78             | 4981       | 126                  | 13.277ms         |

| 14        | 91             | 4949       | 126                  | 15.952ms         |

| 15        | 100            | 4948       | 183                  | 17.642ms         |

| 16        | 100            | 4932       | 199                  | 18.557ms         |

| 17        | 100            | 4914       | 212                  | 19.854ms         |

| 18        | 100            | 4913       | 219                  | 24.856ms         |

| 19        | 100            | 4690       | 221                  | 25.693ms         |

| 20        | 100            | 4681       | 223                  | 26.401ms         |

| 21        | 100            | 4613       | 225                  | 27.211ms         |

| 22        | 100            | 4592       | 227                  | 28.355ms         |

| 23        | 100            | 4568       | 229                  | 29.26ms          |

| 24        | 100            | 4506       | 230                  | 33.81ms          |

| 25        | 100            | 4496       | 233                  | 37.653ms         |

| 26        | 100            | 4398       | 236                  | 38.668ms         |

| 27        | 100            | 4122       | 237                  | 39.631ms         |

| 28        | 100            | 4101       | 237                  | 40.378ms         |

| 29        | 100            | 4074       | 240                  | 45.482ms         |

| 30        | 100            | 4071       | 241                  | 47.503ms         |

| 31        | 100            | 4057       | 241                  | 49.41ms          |

| 32        | 100            | 4056       | 241                  | 51.466ms         |

| 33        | 100            | 4055       | 242                  | 53.697ms         |

| 34        | 100            | 4042       | 243                  | 56.195ms         |

| 35        | 100            | 4014       | 246                  | 59.074ms         |

| 36        | 100            | 3711       | 247                  | 61.956ms         |

| 37        | 100            | 3707       | 250                  | 68.373ms         |

| 38        | 100            | 2661       | 259                  | 72.229ms         |

| 39        | 100            | 2619       | 259                  | 75.719ms         |

| 40        | 100            | 2595       | 259                  | 78.49ms          |

| 41        | 100            | 2593       | 259                  | 81.046ms         |

| 42        | 100            | 2576       | 259                  | 83.581ms         |

| 43        | 100            | 2551       | 259                  | 86.333ms         |

| 44        | 100            | 2544       | 259                  | 89.284ms         |

| 45        | 100            | 2538       | 259                  | 92.283ms         |

| 46        | 100            | 2533       | 259                  | 95.372ms         |

| 47        | 100            | 2530       | 259                  | 98.393ms         |

| 48        | 100            | 2521       | 259                  | 101.47ms         |

| 49        | 100            | 2511       | 259                  | 104.547ms        |

| Final     | 100            | -          | 259                  | 107.66ms         |

+-----------+----------------+------------+----------------------+------------------+

Pattern mining complete. Found 100 unique closed patterns.

Here are the 100 unique closed patterns which are found frequent:

In [19]:
top100_freq_patterns.frequent_patterns.print_rows(num_rows=100, max_column_width=50)

+---------------------------------------------------+---------+
|                      pattern                      | support |
+---------------------------------------------------+---------+
|            [CherryTart, ApricotDanish]            |   2542  |
|           [TuileCookie, MarzipanCookie]           |   2419  |
|          [ChocolateCake, ChocolateCoffee]         |   2176  |
|           [GongolaisCookie, TruffleCake]          |   2131  |
|           [OrangeJuice, CheeseCroissant]          |   2079  |
|              [CherryTart, OperaCake]              |   2067  |
|             [ApricotDanish, OperaCake]            |   2046  |
|         [ApricotCroissant, BlueberryTart]         |   2042  |
|           [StrawberryCake, NapoleonCake]          |   2032  |
|              [CoffeeEclair, ApplePie]             |   1816  |
|            [CoffeeEclair, AlmondTwist]            |   1788  |
|              [ApplePie, AlmondTwist]              |   1785  |
|             [BerryTart, BottledWater] 

Next, we apply the **`extract_features()`** method of this newly trained **`top100_freq_patterns`** model on the **`test`** data set.

In [20]:
features = top100_freq_patterns.extract_features(train)

Preprocessing complete. Found 73566 unique transactions.

Once the features are extracted, we can use them downstream in other applications such as **clustering**, **classification**, **churn prediction**, **recommender systems** etc.

In [23]:
features.print_rows(num_rows=10, max_row_width=120, max_column_width=100)

+---------+----------+
| Receipt | StoreNum |
+---------+----------+
|   8829  |    7     |
|  35020  |    14    |
|  38810  |    20    |
|   6015  |    7     |
|  68521  |    3     |
|   6447  |    5     |
|  12900  |    10    |
|  49658  |    20    |
|  20227  |    19    |
|  47850  |    2     |
+---------+----------+
+-----------------------------------------------------------------------------------------------------+
|                                          extracted_features                                         |
+-----------------------------------------------------------------------------------------------------+
| [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, ... |
| [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ... |
| [0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ... |
| [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0

### Example: Employee Space Clustering by using occurrences of frequency patterns

First we provide an aggregated form of our data by selecting one **Selling Employee (`EmpId`)** at random.

In [24]:
emps = train.groupby(['Receipt', 'StoreNum'],
                     {'EmpId': agg.SELECT_ONE('EmpId')})
emps

Columns:
	Receipt	int
	StoreNum	int
	EmpId	int

Rows: 73566

Data:
+---------+----------+-------+
| Receipt | StoreNum | EmpId |
+---------+----------+-------+
|   8829  |    7     |   7   |
|  35020  |    14    |   14  |
|  38810  |    20    |   20  |
|   6015  |    7     |   7   |
|  68521  |    3     |   3   |
|   6447  |    5     |   5   |
|  12900  |    10    |   35  |
|  49658  |    20    |   20  |
|  20227  |    19    |   50  |
|  47850  |    2     |   23  |
+---------+----------+-------+
[73566 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

Next, we count the instances that each of the **`top100_freq_patterns`** occurs per **`EmpId`**.

In [25]:
emp_space = emps.join(features).groupby('EmpId', {'all_features': agg.SUM('extracted_features')})
emp_space

Columns:
	EmpId	int
	all_features	array

Rows: 50

Data:
+-------+-------------------------------+
| EmpId |          all_features         |
+-------+-------------------------------+
|   49  | [41.0, 53.0, 35.0, 42.0, 4... |
|   13  | [126.0, 130.0, 95.0, 77.0,... |
|   14  | [140.0, 113.0, 107.0, 101.... |
|   48  | [28.0, 34.0, 28.0, 23.0, 2... |
|   20  | [133.0, 129.0, 123.0, 110.... |
|   12  | [15.0, 19.0, 16.0, 14.0, 1... |
|   22  | [52.0, 48.0, 37.0, 45.0, 4... |
|   36  | [35.0, 35.0, 46.0, 38.0, 3... |
|   10  | [38.0, 40.0, 35.0, 40.0, 3... |
|   29  | [17.0, 10.0, 13.0, 19.0, 1... |
+-------+-------------------------------+
[50 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

Finally, we train a **kmeans** algorithm to produce a **3 centered cluster**.

In [26]:
cl_model = gl.kmeans.create(emp_space, 
                            features = ['all_features'], 
                            num_clusters=3)

Batch size is larger than the input dataset. Switching to an exact Kmeans method.

Choosing initial cluster centers with Kmeans++.

+---------------+-----------+

| Center number | Row index |

+---------------+-----------+

| 0             | 43        |

| 1             | 35        |

| 2             | 4         |

+---------------+-----------+

Starting kmeans model training.

Assigning points to initial cluster centers.

+-----------+-------------------------------+

| Iteration | Number of changed assignments |

+-----------+-------------------------------+

| 1         | 0                             |

+-----------+-------------------------------+

In [27]:
emp_space['cluster_id'] = cl_model['cluster_id']['cluster_id']
emp_space

Columns:
	EmpId	int
	all_features	array
	cluster_id	int

Rows: 50

Data:
+-------+-------------------------------+------------+
| EmpId |          all_features         | cluster_id |
+-------+-------------------------------+------------+
|   49  | [41.0, 53.0, 35.0, 42.0, 4... |     1      |
|   13  | [126.0, 130.0, 95.0, 77.0,... |     0      |
|   14  | [140.0, 113.0, 107.0, 101.... |     2      |
|   48  | [28.0, 34.0, 28.0, 23.0, 2... |     1      |
|   20  | [133.0, 129.0, 123.0, 110.... |     2      |
|   12  | [15.0, 19.0, 16.0, 14.0, 1... |     1      |
|   22  | [52.0, 48.0, 37.0, 45.0, 4... |     1      |
|   36  | [35.0, 35.0, 46.0, 38.0, 3... |     1      |
|   10  | [38.0, 40.0, 35.0, 40.0, 3... |     1      |
|   29  | [17.0, 10.0, 13.0, 19.0, 1... |     1      |
+-------+-------------------------------+------------+
[50 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [29]:
gl.canvas.set_target('ipynb')
emp_space.show(x='cluster_id', y = 'StoreNum', view='Bar Chart')